pkg/yamlmeta/node.go

```diff
@@ -69,13 +70,26 @@ func (a *Array) SetValue(val interface{}) error {
 }
 
 func (ai *ArrayItem) SetValue(val interface{}) error {
-	if isMapOrArrayItem(val) {
-		return fmt.Errorf("cannot set map-or-array-item value (%T) into arrayitem", val)
+	if !(val == nil || isMapOrArrayOrScalar(val)) {
+		return fmt.Errorf("arrays can contain only maps, arrays, or scalars; this is a %T", val)
 	}
 	ai.Value = val
 	return nil
 }
 
+func isMapOrArrayOrScalar(val interface{}) bool {
+	switch val.(type) {
+	case *Map, *orderedmap.Map:
```

Comment:
do values actually end up being orderedmap or []interface{} at this level?

Reply:
They do: when nodes originate as Starlark values:

Dict:
https://github.com/vmware-tanzu/carvel-ytt/blob/01e2421ff04a2fa22c634b7cb3eedf04e054d7aa/pkg/template/core/starlark_value.go#L107-L116

List:
https://github.com/vmware-tanzu/carvel-ytt/blob/01e2421ff04a2fa22c634b7cb3eedf04e054d7aa/pkg/template/core/starlark_value.go#L140-L150